experimental/CliffordAlgOrd/src/CliffordAlgebra.jl

```diff
@@ -0,0 +1,563 @@
+      C, K.(coeff)
+    )
+  end
+end
+
+elem_type(::Type{CliffordAlgebra{T, S}}) where {T, S} = CliffordAlgebraElem{T, S}
+
+parent_type(::Type{CliffordAlgebraElem{T, S}}) where {T, S} = CliffordAlgebra{T, S}
+
+base_ring_type(::Type{CliffordAlgebra{T, S}}) where {T, S} = parent_type(T)
+
+################################################################################
+#
+#  Construction
+#
+################################################################################
+
+### Algebra ###
+
+@doc raw"""
+    clifford_algebra(qs::QuadSpace) -> CliffordAlgebra
+
+Return the Clifford algebra of the quadratic space 'qs'.
+"""
+clifford_algebra(qs::Hecke.QuadSpace) =
+  CliffordAlgebra{elem_type(base_ring(qs)),typeof(gram_matrix(qs))}(qs)
+
+### Elements ###
+(C::CliffordAlgebra)() = CliffordAlgebraElem(C)
+
+function (C::CliffordAlgebra)(a::R) where {R<:RingElem}
+  res = fill(zero(a), dim(C))
+  res[1] = a
+  return CliffordAlgebraElem(C, res)
+end
+
+function (C::CliffordAlgebra)(a::R) where {R<:Number}
+  res = fill(zero(a), dim(C))
+  res[1] = a
+  return CliffordAlgebraElem(C, res)
+end
+
+(C::CliffordAlgebra)(coeff::Vector{R}) where {R} = CliffordAlgebraElem(C, coeff)
+
+################################################################################
+#
+#  Basic field access
+#
+################################################################################
+
+### Algebra ###
+@doc raw"""
+    base_ring(C::CliffordAlgebra) -> Ring
+
+Return the base ring of the Clifford algebra $C$.
+"""
+base_ring(C::CliffordAlgebra) = C.base_ring::base_ring_type(typeof(C))
+
+@doc raw"""
+    space(C::CliffordAlgebra) -> QuadSpace
+
+Return the underlying quadratic space of the Clifford algebra $C$.
+"""
+space(C::CliffordAlgebra{T, S}) where {T, S} = C.space::Hecke.QuadSpace{parent_type(T), S}
+
+@doc raw"""
+    gram_matrix(C::CliffordAlgebra) -> MatElem
+
+Return the Gram matrix of the free quadratic module with Clifford algebra $C$.
+"""
+gram_matrix(C::CliffordAlgebra) = C.gram
+
+@doc raw"""
+    dim(C::CliffordAlgebra) -> Int
+
+Return the dimension of the Clifford algebra $C$ over its base ring.
+"""
+dim(C::CliffordAlgebra) = C.dim
+
+### Elements ###
+@doc raw"""
+    parent(x::CliffordAlgebraElem) -> CliffordAlgebra
+
+Return the Clifford algebra containing $x$.
+"""
+parent(x::CliffordAlgebraElem) = x.parent
+
+@doc raw"""
+    coefficients(x::CliffordAlgebraElem) -> Vector
+
+Return the coefficient vector of $x$ wrt the
+canonical basis of its parent Clifford algebra.  
+"""
+coefficients(x::CliffordAlgebraElem) = x.coeffs
+
+@doc raw"""
+    even_coefficients(x::CliffordAlgebraElem) -> Vector
+
+Return the coefficient vector of $x$ wrt the
+canonical basis of its parent Clifford algebra,
+but all coefficients corresponding to basis elements
+with odd grading are set to zero. This also updates
+the field `x.even_coeffs`.
+"""
+function even_coefficients(x::CliffordAlgebraElem)
+  if isdefined(x, :even_coeffs)
+    return x.even_coeffs
```

Comment:
This eg needs a type assertion as well (and 2 times below, and the other functions here)

Reply:
Really? I thought the [type declarations of the data structure](https://github.com/oscar-system/Oscar.jl/pull/4370/files/f88323ea5a13e62d73c1ce5479faa7d82fc88d1a#diff-f830cdbcac3fbb51f8f7c3ae24d7af50985ce58653736e6ae50fc05943170069R45-R47) would be sufficient in this case.

Reply:
Yes they are. I didn't notice that you changed that up there